### PR TITLE
feat: Deep connector Instantly (Write)

### LINKED ISSUE #912: Add Instantly deep connector

### PR DESCRIPTION
**Pull request by Cobalt0s**
Closes #912

# Description
Write endpoints are hard coded to the list of object names associated with them.

There is a second commit with minor tidy things. They are not significant to be in the independent PR but they are not related to current.

# Testing
image

### CODE REVIEW

**Comment by Cobalt0s** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/933?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#933** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/933?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#923** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/923?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#925](https://github.com/amp-labs/connectors/pull/925) <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/925?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#922** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/922?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#914** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/914?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#915](https://github.com/amp-labs/connectors/pull/915) <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/915?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Cobalt0s and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: instantly/read.go [hunk 1]
```diff
@@ -14,6 +14,10 @@ const (
 	objectNameAccounts  = "accounts"
 	objectNameEmails    = "emails"
 	objectNameTags      = "tags"
+
+	objectNameLeads            = "leads"
+	objectNameBlocklistEntries = "blocklist-entries"
+	objectNameUniboxReplies    = "unibox-replies"
```
Comment (jlimatampersand) [inline instantly/read.go:20]: Since these constants are also used in delete and write, we should move to a new file that can be shared across actions.

**Review by jlimatampersand — CHANGES_REQUESTED** [review]
Please resolve comments and re-request review.

File: instantly/write.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+// Documentation links can be found within completeURLPath function.
+func (c *Connector) Write(
+	ctx context.Context, config common.WriteParams,
+) (*common.WriteResult, error) {
+	if len(config.ObjectName) == 0 {
+		return nil, common.ErrMissingObjects
+	}
+
+	// Object name is fully replace
+	url, err := c.getURL()
```
[the remark this answers is not in the document]
Comment (Cobalt0s) [reply instantly/write.go:34]: 100%

File: instantly/write.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+		}, nil
+	}
+
+	// write response was with payload
+	return constructWriteResult(res.Body, recordIdLocation)
+}
+
+// Maps common.WriteParams to URL format for that object.
+// In addition, returns field name where ID will be stored on successful response.
+func completeURLPath(
```
[the remark this answers is not in the document]
Comment (Cobalt0s) [reply instantly/write.go]: I decided to create a map to hold json path to recordIDs.
The same thing I applied to Smartlead.
Instantly: https://github.com/amp-labs/connectors/pull/933/commits/3dc08bba7db8511baea9c9886f56923e62790781
Smartlead: https://github.com/amp-labs/connectors/pull/933/commits/43a9b2e49c4e90376a1b9e7ae2a5eab9e459f90c

PS: Its funny how when I work on multiple deep connectors both of them have shared unique code. Those connectors come in pairs ahaha.

File: providers/instantly/write.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		// Create tag.
+		// https://developer.instantly.ai/tags/create-a-new-tag
+		url.AddPath("custom-tag")
+	}
+}
+
+func constructURLPathUpdate(config common.WriteParams, url *urlbuilder.URL) {
+	if config.ObjectName == objectNameTags {
+		// Update account.
+		// https://developer.instantly.ai/tags/update-tag
```
Comment (jlimatampersand) [inline providers/instantly/write.go:93]: nice, thanks for having all these doc url. These are super helpful.

File: providers/smartlead/objectNames.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+import "github.com/amp-labs/connectors/common/handy"
+
+const (
+	objectNameCampaign     = "campaigns"
+	objectNameEmailAccount = "email-accounts"
+	objectNameClient       = "client"
+)
+
+var supportedObjectsByWrite = handy.NewSet([]string{ //nolint:gochecknoglobals
```
Comment (jlimatampersand) [inline providers/smartlead/objectNames.go:11]: were these only requirements? not lead? I don't have knowledge on the requirements myself.

File: providers/instantly/write.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		url.AddPath("custom-tag", config.RecordId)
+	}
+}
+
+func constructWriteResult(body *ajson.Node, recordIdLocation *string) (*common.WriteResult, error) {
+	if recordIdLocation == nil {
+		// ID is not present inside response. Therefore, empty.
```
Comment (jlimatampersand) [inline providers/instantly/write.go:100]: how do we then return result for lead? we should read lead as well. For lead, email is the unique key they use to identify the object.

File: providers/instantly/write.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	recordIdNodePath := recordIdPaths[config.ObjectName]
+
+	// write response was with payload
+	return constructWriteResult(res.Body, recordIdNodePath)
+}
+
+var recordIdPaths = map[string]*string{ // nolint:gochecknoglobals
+	objectNameLeads:            nil, // ID is not returned for Leads.
```
Comment (jlimatampersand) [inline providers/instantly/write.go:63]: however, emails are used to identify a lead along with campaign id. we should handle this.

**Review by jlimatampersand — CHANGES_REQUESTED** [review]
Overall looks okay except that handling lead object seems to be missing some pieces. Please fill in the gaps or leave some comments on limitations on the lead objects. 

cc @laurenzlong I am not sure about the requirements for this connector, so please comment if missing lead is not an issue for us.

File: providers/instantly/delete.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+		return nil, common.ErrMissingRecordID
+	}
+
+	if !supportedObjectsByDelete.Has(config.ObjectName) {
+		// Removing campaign is the only to be supported at this time.
+		// https://api.smartlead.ai/reference/delete-campaign
+		return nil, common.ErrOperationNotSupportedForObject
+	}
+
+	url, err := c.getURL("custom-tag", config.RecordId)
```
Comment (jlimatampersand) [inline providers/instantly/delete.go:25]: we shouldn't hard code this. I don't think you can delete campaign with custom-tag either. Campaign has id.

File: providers/instantly/delete.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+	if len(config.ObjectName) == 0 {
+		return nil, common.ErrMissingObjects
+	}
+
+	if len(config.RecordId) == 0 {
+		return nil, common.ErrMissingRecordID
+	}
+
+	if !supportedObjectsByDelete.Has(config.ObjectName) {
+		// Removing campaign is the only to be supported at this time.
```
Comment (jlimatampersand) [inline providers/instantly/delete.go:20]: Is this business decision? @laurenzlong? instantly API seems to supprot delete for all objects.

File: providers/smartlead/objectNames.go [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package smartlead
```
Comment (jlimatampersand) [inline providers/smartlead/objectNames.go:1]: this file should be in separate PR.

**Review by jlimatampersand — CHANGES_REQUESTED** [review]

File: providers/instantly/delete.go (hunk 9 above)
Comment (laurenzlong) [reply providers/instantly/delete.go:20]: no, not a business decision, Constantin can you elaborate on the reason for this?

File: providers/instantly/write.go (hunk 6 above)
Comment (Cobalt0s) [reply providers/instantly/write.go:100]: As mentioned above I couldn't explore leads.
I can make `Read leads` request and there is no ID. There is no single read. Only search by email (funny enough only then ID can be returned according to docs if we can believe it, but we cannot use it in any way).

Lead REST API doesn't follow conventions. We could return empty recordID but create custom payload instead to mentione email `common.WriteResult.Data`. But then we would need to document that we do special case. But if we go back whoever makes `/api/v1/lead/add` request will provide `email` for each lead so it defeats the purpose. Hence, write method only returns ACK.

File: providers/instantly/write.go (hunk 7 above)
Comment (Cobalt0s) [reply providers/instantly/write.go:63]: I will close this comment as we are discussing this in the thread above.

File: providers/instantly/delete.go (hunk 8 above)
Comment (Cobalt0s) [reply providers/instantly/delete.go:25]: I messed up the method documentation. I didn't update it.
The delete only works on `tags` which they call `custom-tag`.

It has nothing to do with campaigns. Speaking of campaigns they cannot be removed or created programatically.

File: providers/instantly/delete.go (hunk 9 above)
Comment (Cobalt0s) [reply providers/instantly/delete.go:20]: :sweat: the comment line says campaign while it should say `tags`. Thats a typo.

I created a list of supported objects for Read, Write, Delete bound to current design of inpurt params.
Speaking of Delete I created this connector interface to match others. 

I am not sure about importance of usage of Delete. So let me know if I need to spend time for the following upgrade:
* add `common.DeleteResult.Data` field to store user payload
* modify HTTP client's Delete method

Supporting Delete payload will enable the following Instantly operations:
* Delete Account
* Delete Customer Tracking Domain (this is not a List resource)
* Remove entries from blocklist
* Delete leads from a campaign

PS: my main focus was on Read/Write. This is the first connector to challenge simple OOB Delete. Usually it is a good REST API practice to not have payload during Delete. But APIs are different some may do this via POST, etc.

File: providers/smartlead/objectNames.go (hunk 5 above)
Comment (jlimatampersand) [reply providers/smartlead/objectNames.go:11]: I got confused with instantly because this is instantly PR, so I thought there should be `lead`.

**Review by jlimatampersand — APPROVED** [review]

File: providers/instantly/delete.go (hunk 9 above)
Comment (jlimatampersand) [reply providers/instantly/delete.go]: @laurenzlong so I think main challenge is that delete does not have single id or key reference that can delete objects. It requires campaign id and and email to delete leads. So basically they use composite key like referencing, so we need to have multiple parameters to delete a lead. I think, if delete is not part of requirement we can skip for now. Otherwise, we need to support multi-primary key parameters. 

@Cobalt0s other parts of code look good, but Lauren should make decision on how we want to support delete. I will delegate PR approval to Lauren.

File: providers/smartlead/objectNames.go (hunk 5 above)
Comment (Cobalt0s) [reply providers/smartlead/objectNames.go:11]: Smartlead and Instantly have related set of Objects but with some differences.

File: providers/instantly/delete.go (hunk 9 above)
Comment (Cobalt0s) [reply providers/instantly/delete.go]: Yes, `multi-primary key parameters` is the general rule that I was using to categorize objects for Read, Write, and by inertia I did it ignore Delete.
Luckily, those operations that I listed above can be managed with single ID or no IDs at all. This information can be passed via `Request Body`, but that would require implementation changes I had mentioned. It depends on the business requirements.

Having delete is nice to roll back create/update which was used as a type of cleanup. People using connectors as a library may find it useful. As of now I don't see usage of `Delete` in server.

Lets wait for the roadmap from Lauren.

File: providers/smartlead/objectNames.go (hunk 10 above)
Comment (Cobalt0s) [reply providers/smartlead/objectNames.go:1]: Moved here https://github.com/amp-labs/connectors/pull/1006

File: providers/instantly/delete.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
+package instantly
+
+import (
+	"context"
+
+	"github.com/amp-labs/connectors/common"
+)
+
+// Delete removes object. As of now only removal of Tags are allowed.
```
Comment (laurenzlong) [inline providers/instantly/delete.go:9]:
```suggestion
// Delete removes object. As of now only removal of Tags are allowed because
// deletion of other object types require a request payload to be added, c.Client.Delete does not yet support this.
```

**Review by laurenzlong — APPROVED** [review]

File: providers/instantly/delete.go (hunk 9 above)
Comment (laurenzlong) [reply providers/instantly/delete.go]: We don't yet expose the ability to delete to our customers, so only supporting tags in this connector for now is fine - @Cobalt0s can you take a look at my suggestions on the comment line? Hopefully I captured the information correctly.

File: providers/instantly/delete.go (hunk 11 above)
Comment (Cobalt0s) [reply providers/instantly/delete.go:9]: Slight difference: it uses DELETE with payload. Not POST.
